File: chess.ts/board.ts
import { Color } from "./color"
import { Field } from "./field"
import { Figure } from "./figure"
import { Move, RegularMove, PromotionMove, EnPassantMove, CastlingMove  } from "./move"
import * as U from './util'

export type Board = { [key: string]: Figure }
export const startingBoard: Board = {}
startingBoard[new Field(1,1).show()] = new Figure('Rook','White')
startingBoard[new Field(2,1).show()] = new Figure('Knight','White')
startingBoard[new Field(3,1).show()] = new Figure('Bishop','White')
startingBoard[new Field(4,1).show()] = new Figure('Queen','White')
startingBoard[new Field(5,1).show()] = new Figure('King','White')
startingBoard[new Field(6,1).show()] = new Figure('Bishop','White')
startingBoard[new Field(7,1).show()] = new Figure('Knight','White')
startingBoard[new Field(8,1).show()] = new Figure('Rook','White')
startingBoard[new Field(1,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(2,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(3,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(4,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(5,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(6,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(7,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(8,2).show()] = new Figure('Pawn','White')
startingBoard[new Field(1,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(2,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(3,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(4,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(5,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(6,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(7,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(8,7).show()] = new Figure('Pawn','Black')
startingBoard[new Field(1,8).show()] = new Figure('Rook','Black')
startingBoard[new Field(2,8).show()] = new Figure('Knight','Black')
startingBoard[new Field(3,8).show()] = new Figure('Bishop','Black')
startingBoard[new Field(4,8).show()] = new Figure('Queen','Black')
startingBoard[new Field(5,8).show()] = new Figure('King','Black')
startingBoard[new Field(6,8).show()] = new Figure('Bishop','Black')
startingBoard[new Field(7,8).show()] = new Figure('Knight','Black')
startingBoard[new Field(8,8).show()] = new Figure('Rook','Black')

export function updateBoard(board: Board, move: Move): Board {
    let figure = board[move.from.show()]
    if (figure === undefined) return board;
    let result: Board = Object.assign({}, board);
    if (move instanceof RegularMove) {
        delete result[move.from.show()];
        result[move.to.show()] = figure;
        return result;
    } else if (move instanceof PromotionMove) {
        delete result[move.from.show()];
        result[move.to.show()] = move.figure;
        return result;
    } else if (move instanceof EnPassantMove) {
        delete result[move.from.show()];
        delete result[move.captured.show()];
        result[move.to.show()] = figure;
        return result;
    } else if (move instanceof CastlingMove) {
        let rookFigure = board[move.rookFrom.show()];
        if (rookFigure === undefined) return board;
        delete result[move.from.show()];
        delete result[move.rookFrom.show()];
        result[move.to.show()] = figure;
        result[move.rookTo.show()] = rookFigure;
        return result;
    }
}

// Verifies if the given field is empty
export function isFieldEmpty(board: Board, field: Field): boolean {
    return board[field.show()] === undefined;
}

// Take while condition holds from a list of field lists.
export function takeFieldsWhile(board: Board, fieldss: Field[][], condition: (board: Board, field: Field) => boolean): Field[] {
    return fieldss.flatMap(fields => U.takeWhile(field => condition(board, field), fields))
}


// Returns free fields onto which the figure may be moved
export function freeDestinations(board: Board, fieldss: Field[][]): Field[] {
    return takeFieldsWhile(board, fieldss, isFieldEmpty)
}

// Checks whether the field holds a figure with the given color
export function hasFigureOfColor(board: Board, field: Field, color: Color): boolean {
    let figure = board[field.show()]
    if (figure === undefined) return false;
    else return figure.figureColor === color
}

// Returns fields occupied by the figures of a specific color that are first in each field sequence after an optional series of empty fields
export function findColorFigureFields(board: Board, fieldss: Field[][], color: Color): Field[] {
    return fieldss.flatMap(fields => U.dropWhile(field => isFieldEmpty(board, field), fields).slice(0,1)).filter(f => hasFigureOfColor(board, f, color))
}

File: chess.py/board.py
from itertools import chain, dropwhile, takewhile

from color import Color
from field import Field
from figure import Figure, FigureType
from move import CastlingMove, EnPassantMove, PromotionMove, RegularMove


def starting_board():
    return {
        (1, 1): Figure(FigureType.Rook, Color.White),
        (2, 1): Figure(FigureType.Knight, Color.White),
        (3, 1): Figure(FigureType.Bishop, Color.White),
        (4, 1): Figure(FigureType.Queen, Color.White),
        (5, 1): Figure(FigureType.King, Color.White),
        (6, 1): Figure(FigureType.Bishop, Color.White),
        (7, 1): Figure(FigureType.Knight, Color.White),
        (8, 1): Figure(FigureType.Rook, Color.White),
        (1, 2): Figure(FigureType.Pawn, Color.White),
        (2, 2): Figure(FigureType.Pawn, Color.White),
        (3, 2): Figure(FigureType.Pawn, Color.White),
        (4, 2): Figure(FigureType.Pawn, Color.White),
        (5, 2): Figure(FigureType.Pawn, Color.White),
        (6, 2): Figure(FigureType.Pawn, Color.White),
        (7, 2): Figure(FigureType.Pawn, Color.White),
        (8, 2): Figure(FigureType.Pawn, Color.White),
        (1, 7): Figure(FigureType.Pawn, Color.Black),
        (2, 7): Figure(FigureType.Pawn, Color.Black),
        (3, 7): Figure(FigureType.Pawn, Color.Black),
        (4, 7): Figure(FigureType.Pawn, Color.Black),
        (5, 7): Figure(FigureType.Pawn, Color.Black),
        (6, 7): Figure(FigureType.Pawn, Color.Black),
        (7, 7): Figure(FigureType.Pawn, Color.Black),
        (8, 7): Figure(FigureType.Pawn, Color.Black),
        (1, 8): Figure(FigureType.Rook, Color.Black),
        (2, 8): Figure(FigureType.Knight, Color.Black),
        (3, 8): Figure(FigureType.Bishop, Color.Black),
        (4, 8): Figure(FigureType.Queen, Color.Black),
        (5, 8): Figure(FigureType.King, Color.Black),
        (6, 8): Figure(FigureType.Bishop, Color.Black),
        (7, 8): Figure(FigureType.Knight, Color.Black),
        (8, 8): Figure(FigureType.Rook, Color.Black),
    }


def show_board(board):
    """
    Shows the board.

    >>> print(show_board(starting_board()))
     abcdefgh
    8RNBQKBNR8
    7PPPPPPPP7
    6........6
    5........5
    4........4
    3........3
    2pppppppp2
    1rnbqkbnr1
     abcdefgh
    """

    def row_to_str(row):
        return "".join([str(board.get((col, row), ".")) for col in range(1, 9)])

    return (
        " abcdefgh\n"
        + "".join(
            [(str(row) + row_to_str(row) + str(row) + "\n") for row in range(8, 0, -1)]
        )
        + " abcdefgh"
    )


def update_board(board, move):
    """
    Returns a new board, updated with a move.
    """
    match move:
        case RegularMove(frm, to):
            if frm.as_pair() in board:
                new_board = board.copy()
                new_board[to.as_pair()] = new_board[frm.as_pair()]
                del new_board[frm.as_pair()]
                return new_board
            else:
                return board.copy()
        case PromotionMove(frm, to, figure):
            if frm.as_pair() in board:
                new_board = board.copy()
                new_board[to.as_pair()] = move.figure
                del new_board[frm.as_pair()]
                return new_board
            else:
                return board.copy()
        case EnPassantMove(frm, to, captured):
            if frm.as_pair() in board:
                new_board = board.copy()
                new_board[to.as_pair()] = new_board[frm.as_pair()]
                del new_board[frm.as_pair()]
                del new_board[captured.as_pair()]
                return new_board
            else:
                return board.copy()
        case CastlingMove(frm, to, rook_from, rook_to):
            if frm.as_pair() in board and rook_from.as_pair() in board:
                new_board = board.copy()
                new_board[to.as_pair()] = new_board[frm.as_pair()]
                new_board[rook_to.as_pair()] = new_board[rook_from.as_pair()]
                del new_board[frm.as_pair()]
                del new_board[rook_from.as_pair()]
                return new_board
            else:
                return board.copy()


def is_field_empty(board, field):
    """
    Verifies if the given field is empty.

    >>> is_field_empty(starting_board(), Field(2,2))
    False
    """
    return not (field.as_pair() in board)


def take_fields_while(board, fieldss, condition):
    """
    Take while condition holds from a list of field lists.
    """
    return list(
        chain(
            *[
                list(takewhile(lambda field: condition(board, field), fields))
                for fields in fieldss
            ]
        )
    )


def free_destinations(board, fieldss):
    """
    Returns free fields onto which the figure may be moved
    """
    return take_fields_while(board, fieldss, is_field_empty)


def has_figure_of_color(board, field, color):
    """
    Checks whether the field holds a figure with the given color
    """
    figure = board.get(field.as_pair())
    if figure:
        return figure.figure_color == color
    else:
        return False


def find_color_figure_fields(board, fieldss, color):
    """
    Returns fields occupied by the figures of a specific color that are first in each field sequence after an optional series of empty fields
    """
    return list(
        chain(
            *[
                list(
                    filter(
                        lambda f: has_figure_of_color(board, f, color),
                        list(
                            dropwhile(
                                lambda field: is_field_empty(board, field), fields
                            )
                        )[:1],
                    )
                )
                for fields in fieldss
            ]
        )
    )
